AX: Add tests for RTL/LTR bidi text

gr3g-apple-com opened the pull request:
#### c275e7895a0215704795fd1cbe080265855ab24b
<pre>
AX: Add tests for RTL/LTR bidi text
<a href="https://bugs.webkit.org/show_bug.cgi?id=294460">https://bugs.webkit.org/show_bug.cgi?id=294460</a>
<a href="https://rdar.apple.com/153327595">rdar://153327595</a>

Reviewed by Tyler Wilcock.

Add tests for bidi text: English-Arabic and English-Hebrew. This test
counts the number of LTR and RTL positions within strings based on
textmarkers. There are six strings tested, each with a unique value
applied to its unicode-bidi css property.

* LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic.html: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-hebrew-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-hebrew.html: Added.
* LayoutTests/platform/ios/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/296764@main">https://commits.webkit.org/296764@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/199a68d82054b252cd4d8b14690a34bbbea6cd27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114708 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59718 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29841 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37750 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83231 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112452 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23772 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98635 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63691 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16777 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59324 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93143 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16818 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117822 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27057 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92240 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36916 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92057 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36996 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14742 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32337 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36438 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36101 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39445 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37809 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic.html:
```diff
@@ -0,0 +1,93 @@
+        staticTextTextMarkerRange = staticText.textMarkerRangeForElement(staticText);
+        output += expect("staticText.textMarkerRangeLength(staticTextTextMarkerRange)", "28");
+
+        output += expectRectWithVariance("staticText.boundsForRange(0, 28)", -1, -1, 188, 18, 1);
```
expectRectWithVariance is good, but it still does feel like this test might tend to break if things shift by a couple of pixels.

Would it be possible to instead iterate over the characters and, using the bounds, just assert whether or not each index moved to the left or to the right?

twilco commented on LayoutTests/accessibility/text-marker/text-marker-bidi-element-hebrew.html:
```diff
@@ -0,0 +1,109 @@
+        .unicode-bidi-isolate-override {
+            unicode-bidi: isolate-override;
+        }
+    </style>
```
I think typically in other accessibility layout tests, we don't indent things in the `<head>`.

twilco commented on LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic.html:
```diff
@@ -0,0 +1,108 @@
+        .unicode-bidi-isolate-override {
+            unicode-bidi: isolate-override;
+        }
+    </style>
```
I think typically in other accessibility layout tests, we don't indent things in the `<head>`.

hoffmanjoshua commented on LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic-expected.txt:
```diff
@@ -0,0 +1,196 @@
+
+PASS: rangeLength === 28
+PASS: maxBounds === 188
+PASS: currentBounds > lastBounds === true
```
To better catch failures down the line, I wonder if each pass should print which character is being evaluated, in addition to whether this condition passes?

minorninth replied:
Even better, how about just count the number of characters going each direction.

I'd like something along the lines of:

PASS: characters 0...20 are LTR
PASS: characters 20...26 are RTL
etc.

minorninth commented on LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic-expected.txt:
```diff
@@ -0,0 +1,196 @@
+This test verifies that a combination of Arabic and English is read correctly by Accessibility, when using CSS unicode-bidi property.
+
+PASS: rangeLength === 28
+PASS: maxBounds === 188
```
Maybe delete this line. It looks like it's failing on the bots because we don't get exactly the same number of pixels with and without ITM.

gr3g-apple-com committed:
AX: Add tests for RTL/LTR bidi text
https://bugs.webkit.org/show_bug.cgi?id=294460
rdar://153327595

Reviewed by Tyler Wilcock.

Add tests for bidi text: English-Arabic and English-Hebrew. This test
counts the number of LTR and RTL positions within strings based on
textmarkers. There are six strings tested, each with a unique value
applied to its unicode-bidi css property.

* LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-arabic.html: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-hebrew-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-bidi-element-hebrew.html: Added.
* LayoutTests/platform/ios/TestExpectations:

Canonical link: https://commits.webkit.org/296764@main
